Enchanting Events

APickledWalrus opened the pull request:
### Description
This pull request adds full support for the PrepareItemEnchantEvent and EnchantItemEvent.

---
**Target Minecraft Versions:** Any
**Requirements:**     1.11+ for enchantment offers
**Related Issues:**     None?

Wealthyturtle commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferCost.java:
```diff
@@ -0,0 +1,107 @@
+
+	@Override
+	public @Nullable Class<?>[] acceptChange(Changer.ChangeMode mode) {
+		if (mode == ChangeMode.SET || mode == ChangeMode.ADD || mode == ChangeMode.REMOVE || mode == ChangeMode.RESET)
```
Maybe you could test the other way round, and return Number as default.

Wealthyturtle commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferLevel.java:
```diff
@@ -0,0 +1,126 @@
+
+	@Override
+	public @Nullable Class<?>[] acceptChange(Changer.ChangeMode mode) {
+		if (mode == ChangeMode.SET || mode == ChangeMode.ADD || mode == ChangeMode.REMOVE || mode == ChangeMode.RESET)
```
Same issue here.

Wealthyturtle commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentBonus.java:
```diff
@@ -0,0 +1,87 @@
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (!ScriptLoader.isCurrentEvent(SpongeAbsorbEvent.class)) {
```
Wrong event class xD

APickledWalrus commented:
I have cleaned up the current files and fixed the issues above. I removed the ability to use reset because reset really should change the value back to its original/default value, and that just isn't possible here.

bluelhf commented:
Is it convention to use lower case for types? I personally would prefer PascalCasing

APickledWalrus commented:
The lower case parts will only be used in the code - the display name that you see on the docs would be capitalized.

bensku commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentBonus.java:
```diff
@@ -0,0 +1,81 @@
+import ch.njol.util.Kleenean;
+
+@Name("Enchantment Bonus")
+@Description("The enchantment bonus in an enchant prepare event. This represents the number of bookshelves.")
```
Number of bookshelves where exactly? Near the enchantment table? How near, exactly? You can link to Minecraft wiki if it depends on MC version or is too complex to explain with a few sentences.

bensku commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentExpLevelCosts.java:
```diff
@@ -0,0 +1,160 @@
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Enchantment Exp Level Cost")
+@Description({"The cost of an enchantment in an enchant prepare event. This expression is only for Minecraft 1.9 and 1.10.",
```
Is there any replacement in later versions? Also, use separate annotation to spell out the requirements.

APickledWalrus replied:
There is functionality for this in later versions. It is more advanced though.

bensku commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOffer.java:
```diff
@@ -0,0 +1,91 @@
+import ch.njol.util.Kleenean;
+
+@Name("Enchantment Offer")
+@Description("The enchantment offer in enchant prepare events. Minecraft 1.11+.")
```
Indicate required MC version in annotation meant for requirements.

bensku commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferCost.java:
```diff
@@ -0,0 +1,108 @@
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Enchantment Offer Enchantment Cost")
+@Description({"The cost of an enchantment offer. This is displayed to the right of an enchantment offer. Minecraft 1.11+.",
```
Requirements annotation.

bensku commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferCost.java:
```diff
@@ -0,0 +1,108 @@
+	}
+
+	@Override
+	public @Nullable Class<?>[] acceptChange(Changer.ChangeMode mode) {
```
Nullable annotation to line before.

bensku commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferEnchantment.java:
```diff
@@ -0,0 +1,97 @@
+@Name("Enchantment Offer Enchantment")
+@Description({"The enchantment of an enchantment offer. Minecraft 1.11+.", 
+			"NOTE: The level and cost should be set with their corresponding expressions.", 
+			"This change is visual, and does not effect what the item will be enchanted with.", 
```
Change visual to e.g. cosmetic.

bensku commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferLevel.java:
```diff
@@ -0,0 +1,110 @@
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Enchantment Offer Enchantment Level")
+@Description({"The enchantment level of an enchantment offer. Minecraft 1.11+.",
```
Requirements annotation.

bensku commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferLevel.java:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	@Override
+	public @Nullable Class<?>[] acceptChange(Changer.ChangeMode mode) {
```
No inline nullable annotation.

APickledWalrus commented:
I'd say this PR is pretty much ready. There may be a few changes still necessary (probably documentation), so reviews from anyone who is willing to look through the files would be welcomed. I plan to do a final round of testing, so I will update here when I finish that. 

I have added the enchantment event files, fixed the issues above, and fixed issues with the ``enchantmentoffer`` type.

ShaneBeee commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentBonus.java:
```diff
@@ -0,0 +1,81 @@
+@Name("Enchantment Bonus")
+@Description("The enchantment bonus in an enchant prepare event. This represents the number of bookshelves affecting/surrounding the enchantment table.")
+@Events("enchant prepare")
+@Examples("the enchantment bonus")
```
We often get people asking for help, who demand examples from us (on discord). 
I recommend adding a bit more of an example here, one which could include an event.

APickledWalrus replied:
I’ll make the changes soon! Thanks for the reply :)

APickledWalrus commented:
Changes have been applied. Fixed a few small issues I noticed, and expanded upon the examples.

APickledWalrus commented:
Okay, everything _should_ be good now. Only made a few small changes (full details are in the commit description).

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantItemExpCost.java:
```diff
@@ -0,0 +1,125 @@
+public class ExprEnchantItemExpCost extends SimpleExpression<Number> {
+
+	static {
+		Skript.registerExpression(ExprEnchantItemExpCost.class, Number.class, ExpressionType.SIMPLE, "[the] [exp[erience]] cost of enchant[(ing|ment)]");
```
what about `enchantment cost`? also usually `xp` is valid too

Pikachu920 replied:
or `cost of the enchantment`

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOffer.java:
```diff
@@ -0,0 +1,96 @@
+	static {
+		if (Skript.classExists("org.bukkit.enchantments.EnchantmentOffer")) {
+			Skript.registerExpression(ExprEnchantmentOffer.class, EnchantmentOffer.class, ExpressionType.SIMPLE, 
+					"enchant[ment] offer 1",
```
in my opinion, this expression should not support that singular syntax - you should simply have `enchantment offers` then another expression can handle things like `first element out of enchantment offer`

APickledWalrus replied:
I think it'd be nice to be able to write

`enchantment level of enchantment offer 1`
than
`enchantment level of first element out of enchantment offers`

Pikachu920 replied:
alright, but to be consistent with other expressions the pattern should be `1st enchantment offer` and the number should be an expression otherwise you can't do `enchantment offer {_something}`

https://github.com/SkriptLang/Skript/blob/4c37bc93768db3d0d303c865ca4fd20d29976d44/src/main/java/ch/njol/skript/expressions/ExprLore.java#L70

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOffer.java:
```diff
@@ -0,0 +1,96 @@
+
+	@Override
+	public boolean isSingle() {
+		return offerNumber != 4;
```
wouldn't it be offerNumber != 3?

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferLevel.java:
```diff
@@ -0,0 +1,113 @@
+		if (offers.length == 0)
+			return;
+		int level = delta != null ? ((Number) delta[0]).intValue() : 1;
+		if (level < 1) level = 1;
```
if I try to set it to 0 or a negative number should it really set the level to 1?

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferLevel.java:
```diff
@@ -0,0 +1,113 @@
+			case ADD:
+				for (EnchantmentOffer offer : offers) {
+					change = level + offer.getEnchantmentLevel();
+					if (change < 1)
```
if the change is less than 1 why does it change anything at all?

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferLevel.java:
```diff
@@ -0,0 +1,113 @@
+
+	@Override
+	protected String getPropertyName() {
+		return "[enchant[ment]] level";
```
`getPropertyName` isn't meant to have the syntax tokens in it, it is used for the `toString` and should be representative of how it would read in a script

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferEnchantment.java:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	@Override
+	protected String getPropertyName() {
```
`getPropertyName` isn't meant to have the syntax tokens in it, it is used for the `toString` and should be representative of how it would read in a script

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferCost.java:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	@Override
+	protected String getPropertyName() {
```
`getPropertyName` isn't meant to have the syntax tokens in it, it is used for the `toString` and should be representative of how it would read in a script

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOffer.java:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	@Override
+	public String toString(@Nullable Event e, boolean debug) {
```
doesn't give the right output when the `enchantment offers` expression is used

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentExpLevelCosts.java:
```diff
@@ -0,0 +1,163 @@
+					}
+				} else {
+					subtract = cost - e.getExpLevelCostsOffered()[offerNumber];
+					if (subtract < 1) subtract = 1;
```
subtracting 0 shouldn't subtract 1, it should just do nothing

Pikachu920 commented on src/main/java/ch/njol/skript/classes/data/BukkitClasses.java:
```diff
@@ -1719,6 +1720,41 @@ public String getVariableNamePattern() {
+	
+					@Override
+					public String getVariableNamePattern() {
+						return "\\S+";
```
that implies this will match anything without whitespace which is not true - `toVariableNameString` has a space in it

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantEventsEnchantItem.java:
```diff
@@ -0,0 +1,128 @@
+					e.getItem().setAmount(item.getAmount());
+				}
+				break;
+			case ADD:
```
no need to add these other types

APickledWalrus commented:
I will make some changes soon :)

Thanks for the review!

APickledWalrus commented on src/main/java/ch/njol/skript/expressions/ExprEnchantItemClickedButton.java:
```diff
@@ -0,0 +1,86 @@
+			"\t\tsend \"You clicked button 1!\""})
+@Events("enchant")
+@Since("INSERT VERSION")
+public class ExprEnchantItemClickedButton extends SimpleExpression<Number>{
```
To me, "if the click type is 1" is less clear than "if the clicked enchantment button is 1", but maybe I'm misunderstanding what you mean. Perhaps it would be better to add "clicked enchantment button" to ExprClicked.

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantItemExpCost.java:
```diff
@@ -0,0 +1,127 @@
+
+	static {
+		Skript.registerExpression(ExprEnchantItemExpCost.class, Number.class, ExpressionType.SIMPLE, 
+				"[the] [e]xp[erience] cost (of|for) [the] enchant[ment]",
```
i think it would make more sense to add an enchantment event value to the event and just piggy back off the cost property expression that already exists. it would allow for the same syntax is less code

APickledWalrus replied:
So, this expression is actually the cost that was displayed on the side of the enchantment offer. From what I saw in testing, it doesn't actually do anything if you modify it (I'll test again though), so I'm not why it's even modifiable in the first place. However, since the code is pretty much already there, maybe someone will find a use for being able to modify it. I'll clean up the  code & docs on this one.

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentExpLevelCosts.java:
```diff
@@ -0,0 +1,194 @@
+		if (!Skript.isRunningMinecraft(1, 11)) { // This expression should only be usable on 1.9 and 1.10.
+			Skript.registerExpression(ExprEnchantmentExpLevelCosts.class, Number.class, ExpressionType.SIMPLE,
+					"[the] cost of (enchant[ment]s|enchant[ment] offers)",
+					"[the] cost of enchant[ment] [offer] %number%",
```
instead of a whole expression for this, couldn't you just have this just return an `EnchantmentOffer` and use the already existing cost expression

APickledWalrus replied:
Since EnchantmentOffer wasn't added until 1.11, this expression exists to give users on 1.9 / 1.10 some ability to make changes. The word "offer" in this expression is a little misleading.

Pikachu920 replied:
maybe you could create skript's own EnchantmentOffer and register that on old versions kind of like how ItemType wraps ItemStack and Color wraps Bukkit's color

APickledWalrus replied:
That could be interesting, although this is the only customizable part on 1.9 / 1.10 I believe.

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOffer.java:
```diff
@@ -0,0 +1,151 @@
+		if (multiple)
+			return ((PrepareItemEnchantEvent) e).getOffers();
+
+		int offerNumber = exprOfferNumber.getSingle(e).intValue();
```
`exprOfferNumber` may be null, this could cause an NPE

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOffer.java:
```diff
@@ -0,0 +1,151 @@
+
+		int offerNumber = exprOfferNumber.getSingle(e).intValue();
+		if (offerNumber < 1 || offerNumber > 3)
+			return new EnchantmentOffer[]{};
```
new EnchantmentOffer[0];

Pikachu920 replied:
also wouldn't it make more sense to just check the length of the offer array? just because this is the length now doesn't mean it will be in the future

APickledWalrus replied:
I will have to change that. There could be 0-3 offers, so I'll have to account for that.

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferLevel.java:
```diff
@@ -0,0 +1,114 @@
+			case ADD:
+				for (EnchantmentOffer offer : offers) {
+					change = level + offer.getEnchantmentLevel();
+					if (change < 1) 
```
wouldn't this mean you can't do `add -1 to player's level` to a level 1 player and get 0?

APickledWalrus replied:
This check is to just make sure the level isn't being set to a number < 1, because you aren't able to do that. However, I will add some comments for clarity.

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferLevel.java:
```diff
@@ -0,0 +1,114 @@
+				break;
+			case REMOVE:
+				for (EnchantmentOffer offer : offers) {
+					change = level - offer.getEnchantmentLevel();
```
shouldn't this be reversed? `offer.getEnchantmentLevel() - level`

APickledWalrus replied:
This check is to just make sure the level isn't being set to a number < 1, because you aren't able to do that. However, I will add some comments for clarity.

Pikachu920 commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentOfferLevel.java:
```diff
@@ -0,0 +1,114 @@
+		if (offers.length == 0 || delta == null)
+			return;
+		int level = ((Number) delta[0]).intValue();
+		if (level < 1) 
```
shouldn't scripters be able to set the level to 0 or even negative levels if they wanted to?

APickledWalrus replied:
It does not work - an error is thrown.

Pikachu920 replied:
even if it's set to 0?

APickledWalrus replied:
Yep. I thought it was possible to get enchantments of those levels, but maybe not? The cost and enchantment level both have to be greater than 0.

SnowPyon commented on .classpath:
```diff
@@ -24,7 +24,6 @@
 			<attribute name="gradle_used_by_scope" value="test"/>
```
Remove this change

SnowPyon commented on src/main/java/ch/njol/skript/classes/data/BukkitClasses.java:
```diff
@@ -1719,6 +1720,41 @@ public String getVariableNamePattern() {
+	
+					@Override
+					public String toVariableNameString(EnchantmentOffer eo) {
+						return toString(eo, 0);
```
This should be `"offer:" + eo.getEnchantment().getKey().getKey() + "=" + eo.getEnchantmentLevel()` or something similar, and the varible pattern may match that as well.

SnowPyon commented on src/main/java/ch/njol/skript/expressions/ExprEnchantEventsEnchantItem.java:
```diff
@@ -0,0 +1,129 @@
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Enchant Events Enchant Item")
```
This should just be called `Enchant Item`

SnowPyon commented on src/main/java/ch/njol/skript/expressions/ExprEnchantEventsEnchantItem.java:
```diff
@@ -0,0 +1,129 @@
+			"\tset the enchanted item to a diamond chestplate",
+			"on enchant prepare:",
+			"\tset the enchant item to a wooden sword"})
+@Events("enchant prepare and enchant")
```
This method accepts an array, and it should match the given `name` parameter on Skript#registerEvent method for the event.

SnowPyon commented on src/main/java/ch/njol/skript/expressions/ExprEnchantItemEnchantments.java:
```diff
@@ -0,0 +1,137 @@
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Enchant Event Applied Enchantments")
```
Same as above, remove `Enchant Event` and make @Events match with the event name.

SnowPyon commented on src/main/java/ch/njol/skript/expressions/ExprEnchantItemEnchantments.java:
```diff
@@ -0,0 +1,137 @@
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.REMOVE_ALL || mode == ChangeMode.RESET)
+			return null;
+		return CollectionUtils.array(Enchantment[].class, EnchantmentType[].class);
```
I believe this should be fine if it takes only EnchantmentType due to the Enchantment -> EnchantmentType converter (and Skript shouldn't really be returning Enchantment anywhere, if it does, it should be changed), not 100% sure though.

APickledWalrus replied:
I based this off of the following:
```
// Enchantment doesn't get automatically converted to EnchantmentType if you give it more than a one.
// Meaning you can transform an Enchantment array to an EnchantmentType array automatically,
// So, we gotta do it manually.
return CollectionUtils.array(Enchantment[].class, EnchantmentType[].class);
```
Source: https://github.com/SkriptLang/Skript/blob/14f267cc4bb8f00fa5c13536ce8a80afe64b1465/src/main/java/ch/njol/skript/expressions/ExprEnchantments.java#L89

SnowPyon replied:
Oh right, I wrote that lol. Yeah, we should fix that but for now, just keep accepting both. Sorry for the confusion.

SnowPyon commented on src/main/java/ch/njol/skript/expressions/ExprClicked.java:
```diff
@@ -72,7 +73,8 @@
 		TYPE(4, ClickType.class, "click type", "click (type|action)"),
-		ACTION(5, InventoryAction.class, "inventory action", "inventory action");
+		ACTION(5, InventoryAction.class, "inventory action", "inventory action"),
+		ENCHANT_BUTTON(6, Number.class, "clicked enchantment button", "(clicked [enchant[ment]] button|enchant[ment] button clicked)");
```
`enchantment button clicked` doesn't really make sense, does it?

APickledWalrus replied:
Since it only returns a number, probably not?
`if the enchantment button clicked is 1` sounds okay to me, although it is a bit odd.

SnowPyon commented on src/main/java/ch/njol/skript/expressions/ExprEnchantItemEnchantments.java:
```diff
@@ -0,0 +1,137 @@
+public class ExprEnchantItemEnchantments extends SimpleExpression<EnchantmentType> {
+
+	static {
+		Skript.registerExpression(ExprEnchantItemEnchantments.class, EnchantmentType.class, ExpressionType.SIMPLE, "[the] applied enchant[ment][s]");
```
The `s` shouldn't be optional,

APickledWalrus replied:
I thought since there might only be one enchantment returned I should keep it optional, but since there isn't any way for me to check that in isSingle, I'll make the change

SnowPyon commented on src/main/java/ch/njol/skript/expressions/ExprEnchantItemExpCost.java:
```diff
@@ -0,0 +1,120 @@
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Enchant Event Experience Cost")
```
Same as the others.

SnowPyon commented on src/main/java/ch/njol/skript/expressions/ExprEnchantItemExpCost.java:
```diff
@@ -0,0 +1,120 @@
+
+	static {
+		Skript.registerExpression(ExprEnchantItemExpCost.class, Number.class, ExpressionType.SIMPLE, 
+				"[the] [displayed] [[e]xp[erience]] cost (of enchanting|to enchant)");
```
Remove the `enchanting|to enchant` part, it's redundant.

SnowPyon commented on src/main/java/ch/njol/skript/expressions/ExprEnchantmentBonus.java:
```diff
@@ -0,0 +1,83 @@
+import ch.njol.skript.log.ErrorQuality;
+import ch.njol.util.Kleenean;
+
+@Name("Enchant Prepare Enchantment Bonus")
```
Same here.

APickledWalrus commented:
This PR *should* be ready now too! Thanks to everyone that helped 😄
